pkg/vendir/fetch/hg/hg.go

```diff
@@ -21,12 +21,28 @@ type Hg struct {
 	opts       ctlconf.DirectoryContentsHg
 	infoLog    io.Writer
 	refFetcher ctlfetch.RefFetcher
+	authDir    string
+	env        []string
+	cacheID    string
 }
 
 func NewHg(opts ctlconf.DirectoryContentsHg,
-	infoLog io.Writer, refFetcher ctlfetch.RefFetcher) *Hg {
+	infoLog io.Writer, refFetcher ctlfetch.RefFetcher,
+	tempArea ctlfetch.TempArea,
+) (*Hg, error) {
+	t := Hg{opts, infoLog, refFetcher, "", nil, ""}
+	if err := t.setup(tempArea); err != nil {
+		return nil, err
+	}
+	return &t, nil
+}
 
-	return &Hg{opts, infoLog, refFetcher}
+// CacheID returns a cache id for the repository
+// It doesn't include the ref because we want to reuse a cache when only the ref
+// is changed
+// Basically we combine all data used to write the hgrc file
+func (t *Hg) CacheID() string {
```

Comment:
Changes looks good to me but as being not an expert in hg having low confidence on what all data this cacheID will contain. 

Here we are saying  it won't include the ref (from my understanding could be tag, branch and commit id) and in case of commit id, it will change only  if there is content change. And in this case we will not want to use from cache. 
I am bit confuse how we are saying that the target what we want is same as what we get from cache if we are not storing ref in cache?

Or we are considering this for only tag and branch?

Reply:
The ref can be a changeset id, a tag, a branch, a bookmark, a topic: anything that can designate a mercurial changeset.

No checkout is done on the cached repository, so it is not pointing any particular changeset (in git terms, it has no HEAD).
The goal is to reuse this repository as a base whenever we sync any changeset from the same source URL.

That said the cache ID I wrote had 2 issues:
- it did not contain the URL
- it leaked credentials

I fixed those two issues in the next changeset.